LEGO1/mxwavepresenter.cpp

```diff
@@ -159,63 +194,182 @@ void MxWavePresenter::StartingTickle()
+	if (IsEnabled())
+		m_subscriber->FUN_100b8390(p_chunk);
 }
 
-// OFFSET: LEGO1 0x100b2160 STUB
+// OFFSET: LEGO1 0x100b2160
 undefined4 MxWavePresenter::PutData()
 {
-	// TODO
+	MxAutoLocker lock(&m_criticalSection);
+
+	if (IsEnabled()) {
+		switch (m_currentTickleState) {
+		case TickleState_Streaming:
+			if (m_currentChunk && FUN_100b1ba0()) {
+				WriteToSoundBuffer(m_currentChunk->GetData(), m_currentChunk->GetLength());
+				m_subscriber->FUN_100b8390(m_currentChunk);
+				m_currentChunk = NULL;
+			}
+
+			if (!m_playing) {
+				m_dsBuffer->SetCurrentPosition(0);
+
+				if (m_dsBuffer->Play(0, 0, DSBPLAY_LOOPING) == DS_OK)
+					m_playing = TRUE;
+			}
+			break;
+		case TickleState_Repeating:
+			if (m_playing)
+				break;
+
+			m_dsBuffer->SetCurrentPosition(0);
+
+			if (m_dsBuffer->Play(0, 0, m_action->GetLoopCount() > 1) == DS_OK)
+				m_playing = TRUE;
+		}
+	}
+
 	return 0;
 }
 
-// OFFSET: LEGO1 0x100b2280 STUB
+// OFFSET: LEGO1 0x100b2280
 void MxWavePresenter::EndAction()
 {
-	// TODO
+	if (m_action) {
+		MxAutoLocker lock(&m_criticalSection);
+		MxMediaPresenter::EndAction();
+
+		if (m_dsBuffer)
+			m_dsBuffer->Stop();
+	}
 }
 
-// OFFSET: LEGO1 0x100b2300 STUB
-void MxWavePresenter::SetVolume(MxU32 p_volume)
+// OFFSET: LEGO1 0x100b2300
+void MxWavePresenter::SetVolume(MxS32 p_volume)
 {
-	// TODO
+	m_criticalSection.Enter();
+
+	m_volume = p_volume;
+	if (m_dsBuffer != NULL) {
+		MxS32 volume = p_volume * MxOmni::GetInstance()->GetSoundManager()->GetVolume() / 100;
+		MxS32 otherVolume = MxOmni::GetInstance()->GetSoundManager()->FUN_100aecf0(volume);
+		m_dsBuffer->SetVolume(otherVolume);
+	}
+
+	m_criticalSection.Leave();
 }
 
-// OFFSET: LEGO1 0x100b2360 STUB
+// OFFSET: LEGO1 0x100b2360
 void MxWavePresenter::Enable(MxBool p_enable)
 {
-	// TODO
+	if (IsEnabled() != p_enable) {
+		MxSoundPresenter::Enable(p_enable);
+
+		if (p_enable) {
+			m_writtenChunks = 0;
+			m_playing = FALSE;
+		}
+		else if (m_dsBuffer)
+			m_dsBuffer->Stop();
+	}
 }
 
-// OFFSET: LEGO1 0x100b23a0 STUB
+// OFFSET: LEGO1 0x100b23a0
 void MxWavePresenter::ParseExtra()
 {
-	// TODO
+	char extraCopy[512];
+
+	MxSoundPresenter::ParseExtra();
+	*((MxU16*) &extraCopy[0]) = m_action->GetExtraLength();
+	char* extraData = m_action->GetExtraData();
+
+	if (*((MxU16*) &extraCopy[0])) {
+		MxU16 len = *((MxU16*) &extraCopy[0]);
+		memcpy(extraCopy, extraData, len);
+		extraCopy[len] = '\0';
+
+		char t_soundValue[512];
+		if (KeyValueStringParse(t_soundValue, g_strSOUND, extraCopy)) {
+			if (!strcmpi(t_soundValue, "FALSE"))
+				Enable(FALSE);
+		}
+	}
 }
 
-// OFFSET: LEGO1 0x100b2440 STUB
+// OFFSET: LEGO1 0x100b2440
 void MxWavePresenter::VTable0x64()
 {
-	// TODO
+	if (!m_unk68 && m_playing) {
+		if (m_dsBuffer)
+			m_dsBuffer->Stop();
+		m_unk68 = TRUE;
+	}
 }
 
-// OFFSET: LEGO1 0x100b2470 STUB
+// OFFSET: LEGO1 0x100b2470
 void MxWavePresenter::VTable0x68()
 {
-	// TODO
+	if (m_unk68) {
```

Comment:
Last comment I have is maybe this is just a variable to mark if the buffer is available or not in use

Reply:
Looks more like a stopped/paused flag to me.